mypy/stubtest.py

```diff
@@ -369,25 +358,29 @@ def verify_mypyfile(
         if not o.module_hidden and (not is_probably_private(m) or hasattr(runtime, m))
     }
 
-    imported_symbols = _get_imported_symbol_names(runtime)
-
     def _belongs_to_runtime(r: types.ModuleType, attr: str) -> bool:
         """Heuristics to determine whether a name originates from another module."""
         obj = getattr(r, attr)
         if isinstance(obj, types.ModuleType):
             return False
-        if callable(obj):
-            # It's highly likely to be a class or a function if it's callable,
-            # so the __module__ attribute will give a good indication of which module it comes from
-            try:
-                obj_mod = obj.__module__
-            except Exception:
-                pass
-            else:
-                if isinstance(obj_mod, str):
-                    return bool(obj_mod == r.__name__)
-        if imported_symbols is not None:
-            return attr not in imported_symbols
+
+        symbols = _module_symbol_table(r)
+        if symbols is not None:
+            if attr in {sym.get_name() for sym in symbols.get_symbols() if sym.is_imported()}:
+                # symtable says we got this from another module
+                return False
+
+            # but we can't just return True here, because symtable doesn't know about star imports
+            if attr in {sym.get_name() for sym in symbols.get_symbols() if sym.is_assigned()}:
+                # symtable knows we assigned this symbol in the module
+                return True
```

Comment:
nit: we're now just throwing away these sets as soon as we create them; we could possibly do some short-circuiting here:

```suggestion
            for sym in symbols.get_symbols():
                if sym.is_imported() and attr == sym.get_name():
                    # symtable says we got this from another module
                    return False

            # but we can't just return True here, because symtable doesn't know about star imports
            for sym in symbols.get_symbols():
                if sym.is_assigned() and attr == sym.get_name():
                    # symtable knows we assigned this symbol in the module
                    return True
```